### PR TITLE
Autosaves the overall comments every 3 seconds if there are changes

### LINKED ISSUE #2786: The "Overall Comments" are not autosaved
In the assignment grading view, the marks are autosaved, implying that all changes should be autosaved. But the "Overall Comments" under the annotation tab is not autosaved unless "Save Changes" is clicked. I lost all my comments...

### PR DESCRIPTION
**Pull request by ihasan98**
## Motivation and Context

resolves #2786

## Your Changes

**Description**:

**Type of change** (select all that apply):

 

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring (internal change to codebase, without changing functionality)
- [ ] Test update (change that modifies or updates tests only)
- [ ] Other (please specify): 

## Testing

Manual testing through web interface

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have fixed any Hound bot comments. 
- [x] I have verified that the TravisCI tests have passed.

### CODE REVIEW

File: app/assets/javascripts/Components/Result/annotation_panel.jsx [hunk 1]
```diff
@@ -10,20 +10,22 @@ export class AnnotationPanel extends React.Component {
       overallComment: props.overallComment,
       unsavedChanges: false
     };
+    this.autoSaveOverallComments = this.autoSaveOverallComments.bind(this);
```
Comment (mishaschwartz) [inline app/assets/javascripts/Components/Result/annotation_panel.jsx:13]: Instead of using the "bind in constructor" technique, can we use the "class properties" syntax instead:

https://reactjs.org/docs/faq-functions.html#class-properties-stage-3-proposal

since we're using this syntax in a lot of other react components in MarkUs.

File: app/assets/javascripts/Components/Result/autosave_text_form.jsx [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    timeout = setTimeout(later, wait);
+
+    if (callNow) func.apply(context, args);
+  };
+}
+
+const SaveMessage = ({ unSaved }) => (
+  <div className="autosave-text">
+    {unSaved ? (
+      <p> There are unsaved changes </p>
```
Comment (david-yz-liu) [inline app/assets/javascripts/Components/Result/autosave_text_form.jsx:29]: Please internationalize these strings (see https://guides.rubyonrails.org/i18n.html). You can put the strings under the `views/results/` locale file. It's okay to use English only.

File: app/assets/javascripts/Components/Result/annotation_panel.jsx [hunk 3]
```diff
@@ -1,91 +1,45 @@
...
+      data: { result: { overall_comment: value } },
     });
-    event.preventDefault();
   };
 
   render() {
     let overallCommentElement;
     if (this.props.released_to_students || this.props.remarkSubmitted) {
-      overallCommentElement = <div id='overall_comment_text' key='overall_comment_text' />;
+      overallCommentElement = (
```
Comment (david-yz-liu) [inline app/assets/javascripts/Components/Result/annotation_panel.jsx:20]: Revert this change

File: app/assets/javascripts/Components/Result/annotation_panel.jsx [hunk 4]
```diff
@@ -1,91 +1,45 @@
...
-          <div id="overall_comment_preview" className="preview"/>
-        </div>
+        <TextForm
+          initialValue={this.props.overallComment}
+          persistChanges={this.persistChanges}
+          previewId={"overall_comment_preview"}
+        />
       );
     }
-
```
Comment (david-yz-liu) [inline app/assets/javascripts/Components/Result/annotation_panel.jsx:83]: Please revert the changes here and below

File: app/assets/javascripts/Components/Result/autosave_text_form.jsx [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    }
+  }
+
+  handlePersist = debounce(
+    () => {
+      this.props.persistChanges(this.state.value).then(() => {
+        this.setState({ unsavedChanges: false });
+      });
+    },
+    1500,
```
Comment (david-yz-liu) [inline app/assets/javascripts/Components/Result/autosave_text_form.jsx:65]: I'm wondering what the relationship is between this number and the 3000. You should pull these out into constants and include a comment explaining their purpose.

**Review by david-yz-liu — COMMENTED** [review]
Nice work, @ihasan98. In addition to the few comments I left, please also look into removing the flash message, and making the "success" message appear after a success response. Also the styling for the autosave text could be improved to distinguish between the two messages, maybe including the checkmark icon after "success" message?

File: app/assets/javascripts/Components/Result/autosave_text_form.jsx (hunk 5 above)
Comment (ihasan98) [reply app/assets/javascripts/Components/Result/autosave_text_form.jsx:65]: This was my bad. I forgot to push the commit that removed the `setInterval`. I'll still make this a constant

**Comment by david-yz-liu** [comment]
@ihasan98 I didn't notice this earlier, but currently the overall comment no longer appears when the result is released. The way you've implemented the `TextForm`'s `updatePreview` method isn't correctly rendering to the overall comment text div when the result is released (it's a different one than the preview).
